homeassistant/components/climate/generic_thermostat.py

```diff
@@ -314,62 +305,49 @@ def _async_update_temp(self, state):
-                self.min_cycle_duration)
-            if not long_enough:
+        async with self._temp_lock:
+            if not self._active and None not in (self._cur_temp,
+                                                 self._target_temp):
+                self._active = True
+                _LOGGER.info("Obtained current and target temperature. "
+                             "Generic thermostat active. %s, %s",
+                             self._cur_temp, self._target_temp)
+
+            if not self._active or not self._enabled:
                 return
 
-        if self.ac_mode:
-            is_cooling = self._is_device_active
-            if is_cooling:
-                too_cold = self._target_temp - self._cur_temp >= \
-                    self._cold_tolerance
-                if too_cold:
-                    _LOGGER.info("Turning off AC %s", self.heater_entity_id)
-                    self._heater_turn_off()
-            else:
-                too_hot = self._cur_temp - self._target_temp >= \
-                    self._hot_tolerance
-                if too_hot:
-                    _LOGGER.info("Turning on AC %s", self.heater_entity_id)
-                    self._heater_turn_on()
-        else:
-            is_heating = self._is_device_active
-            if is_heating:
-                too_hot = self._cur_temp - self._target_temp >= \
-                    self._hot_tolerance
-                if too_hot:
+            if self.min_cycle_duration:
+                if self._is_device_active:
+                    current_state = STATE_ON
+                else:
+                    current_state = STATE_OFF
+                long_enough = condition.state(
+                    self.hass, self.heater_entity_id, current_state,
+                    self.min_cycle_duration)
+                if not long_enough:
+                    return
+
+            too_cold = \
+                self._target_temp - self._cur_temp >= self._cold_tolerance
+            too_hot = \
+                self._cur_temp - self._target_temp >= self._hot_tolerance
+            if self._is_device_active:
+                if (self.ac_mode and too_cold) or \
+                   (not self.ac_mode and too_hot):
                     _LOGGER.info("Turning off heater %s",
                                  self.heater_entity_id)
-                    self._heater_turn_off()
+                    await self._async_heater_turn_off()
+                elif time is not None:
+                    await self._async_heater_turn_on()
             else:
-                too_cold = self._target_temp - self._cur_temp >= \
-                    self._cold_tolerance
-                if too_cold:
+                if (self.ac_mode and too_hot) or \
+                   (not self.ac_mode and too_cold):
                     _LOGGER.info("Turning on heater %s", self.heater_entity_id)
-                    self._heater_turn_on()
+                    await self._async_heater_turn_on()
```

Comment:
You have removed the logging that mentions AC and I think that is okay because the configuration for the entity_id is indeed named `heater`, even in `ac_mode`.

Reply:
Yup. I hope we can come up with a better, more neutral naming convention (as heater is very, well, heating-oriented :) ), and when we do, I'll update the logging accordingly.